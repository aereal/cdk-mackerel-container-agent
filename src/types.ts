export enum MackerelContainerPlatform {
  KUBERNETES = "kubernetes", // https://mackerel.io/ja/docs/entry/howto/install-agent/container/kubernetes
  ECS = "ecs" // https://mackerel.io/ja/docs/entry/howto/install-agent/container/ecs
}

export interface ServiceRole {
  role: string;
  service: string;
}

export enum MackerelHostStatus {
  Working = "working",
  Standby = "standby",
  Maintenance = "maintenance",
  Poweroff = "poweroff"
}
