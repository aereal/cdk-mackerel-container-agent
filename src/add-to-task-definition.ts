import {
  Compatibility,
  ContainerDefinition,
  ContainerImage,
  NetworkMode,
  TaskDefinition,
} from "@aws-cdk/aws-ecs"
import {
  AdditionalContainerDefinitionOptions,
  MackerelContainerPlatform,
  ServiceRole,
} from "./types"

interface Props {
  taskDefinition: TaskDefinition
  apiKey: string
  roles?: ReadonlyArray<ServiceRole>
  ignoreContainer?: string
  additionalContainerOptions?: AdditionalContainerDefinitionOptions
}

const guessPlatform = (
  taskDefinition: TaskDefinition
): MackerelContainerPlatform | undefined => {
  switch (taskDefinition.networkMode) {
    case NetworkMode.AwsVpc:
      return taskDefinition.compatibility === Compatibility.Ec2
        ? MackerelContainerPlatform.AWSVPC
        : MackerelContainerPlatform.FARGATE
    case NetworkMode.Bridge:
    case NetworkMode.None:
    case NetworkMode.Host:
      return MackerelContainerPlatform.ECS
    default:
      return undefined
  }
}

export const addMackerelContainerAgent = (
  props: Props
): ContainerDefinition => {
  const {
    apiKey,
    taskDefinition,
    roles,
    ignoreContainer,
    additionalContainerOptions: opts,
  } = props
  const guessedPlatform = guessPlatform(taskDefinition)
  if (!guessedPlatform) {
    throw new Error("Cannot guess platform from taskDefinition")
  }

  const environment: Record<string, string> = {
    ...(opts && opts.environment ? opts.environment : {}),
    MACKEREL_APIKEY: apiKey,
    MACKEREL_CONTAINER_PLATFORM: guessedPlatform,
  }
  if (roles) {
    environment.MACKEREL_ROLES = roles
      .map(({ role, service }) => `${service}:${role}`)
      .join(",")
  }
  if (ignoreContainer) {
    environment.MACKEREL_IGNORE_CONTAINER = ignoreContainer
  }
  const mackerelAgentContainer = taskDefinition.addContainer("mackerel-agent", {
    ...opts,
    environment,
    image: ContainerImage.fromDockerHub(
      "mackerel/mackerel-container-agent:latest"
    ),
    memoryLimitMiB: 128,
  })
  taskDefinition.addVolume({
    host: {
      sourcePath: "/cgroup", // TODO: support AL2
    },
    name: "cgroup",
  })
  mackerelAgentContainer.addMountPoints({
    containerPath: "/host/sys/fs/cgroup",
    readOnly: true,
    sourceVolume: "cgroup",
  })
  taskDefinition.addVolume({
    host: {
      sourcePath: "/var/run/docker.sock",
    },
    name: "docker_sock",
  })
  mackerelAgentContainer.addMountPoints({
    containerPath: "/var/run/docker.sock",
    readOnly: true,
    sourceVolume: "docker_sock",
  })
  return mackerelAgentContainer
}
