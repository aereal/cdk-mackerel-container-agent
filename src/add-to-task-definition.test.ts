import { Ec2TaskDefinition } from "@aws-cdk/aws-ecs"
import { LogGroup } from "@aws-cdk/aws-logs"
import { Stack } from "@aws-cdk/cdk"
import { addMackerelContainerAgent } from "./add-to-task-definition"
import { MackerelContainerPlatform } from "./types"

describe("addMackerelContainerAgent", () => {
  test("add to task definition with only required props", () => {
    const stack = new Stack()
    const taskDefinition = new Ec2TaskDefinition(stack, "TaskDefinition", {})
    const container = addMackerelContainerAgent({
      apiKey: "keep-my-secret",
      taskDefinition,
    })
    expect(stack.toCloudFormation()).toMatchSnapshot()
  })

  test("with roles", () => {
    const stack = new Stack()
    const taskDefinition = new Ec2TaskDefinition(stack, "TaskDefinition", {})
    const container = addMackerelContainerAgent({
      apiKey: "keep-my-secret",
      roles: [
        { service: "My-service", role: "db" },
        { service: "My-service", role: "proxy" },
      ],
      taskDefinition,
    })
    expect(stack.toCloudFormation()).toMatchSnapshot()
  })

  test("specify ignoreContainer", () => {
    const stack = new Stack()
    const taskDefinition = new Ec2TaskDefinition(stack, "TaskDefinition", {})
    const container = addMackerelContainerAgent({
      apiKey: "keep-my-secret",
      ignoreContainer: "(mackerel|xray)",
      taskDefinition,
    })
    expect(stack.toCloudFormation()).toMatchSnapshot()
  })

  test("specify logGroup", () => {
    const stack = new Stack()
    const taskDefinition = new Ec2TaskDefinition(stack, "TaskDefinition", {})
    const container = addMackerelContainerAgent({
      apiKey: "keep-my-secret",
      logGroup: new LogGroup(stack, "LogGroup"),
      taskDefinition,
    })
    expect(stack.toCloudFormation()).toMatchSnapshot()
  })
})
