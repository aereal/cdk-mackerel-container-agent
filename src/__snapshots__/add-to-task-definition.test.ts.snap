// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addMackerelContainerAgent add to task definition with only required props 1`] = `
Object {
  "Resources": Object {
    "TaskDefinitionB36D86D9": Object {
      "Properties": Object {
        "ContainerDefinitions": Array [
          Object {
            "Environment": Array [
              Object {
                "Name": "MACKEREL_APIKEY",
                "Value": "keep-my-secret",
              },
              Object {
                "Name": "MACKEREL_CONTAINER_PLATFORM",
                "Value": "ecs",
              },
            ],
            "Essential": false,
            "Image": "mackerel/mackerel-container-agent:latest",
            "Links": Array [],
            "LinuxParameters": Object {
              "Capabilities": Object {
                "Add": Array [],
                "Drop": Array [],
              },
              "Devices": Array [],
              "Tmpfs": Array [],
            },
            "MemoryReservation": 128,
            "MountPoints": Array [
              Object {
                "ContainerPath": "/host/sys/fs/cgroup",
                "ReadOnly": true,
                "SourceVolume": "cgroup",
              },
              Object {
                "ContainerPath": "/var/run/docker.sock",
                "ReadOnly": true,
                "SourceVolume": "docker_sock",
              },
            ],
            "Name": "mackerel-agent",
            "PortMappings": Array [],
            "Ulimits": Array [],
            "VolumesFrom": Array [],
          },
        ],
        "Family": "TaskDefinition",
        "NetworkMode": "bridge",
        "PlacementConstraints": Array [],
        "RequiresCompatibilities": Array [
          "EC2",
        ],
        "TaskRoleArn": Object {
          "Fn::GetAtt": Array [
            "TaskDefinitionTaskRoleFD40A61D",
            "Arn",
          ],
        },
        "Volumes": Array [
          Object {
            "Host": Object {
              "SourcePath": "/cgroup",
            },
            "Name": "cgroup",
          },
          Object {
            "Host": Object {
              "SourcePath": "/var/run/docker.sock",
            },
            "Name": "docker_sock",
          },
        ],
      },
      "Type": "AWS::ECS::TaskDefinition",
    },
    "TaskDefinitionTaskRoleFD40A61D": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;

exports[`addMackerelContainerAgent specify ignoreContainer 1`] = `
Object {
  "Resources": Object {
    "TaskDefinitionB36D86D9": Object {
      "Properties": Object {
        "ContainerDefinitions": Array [
          Object {
            "Environment": Array [
              Object {
                "Name": "MACKEREL_APIKEY",
                "Value": "keep-my-secret",
              },
              Object {
                "Name": "MACKEREL_CONTAINER_PLATFORM",
                "Value": "ecs",
              },
              Object {
                "Name": "MACKEREL_IGNORE_CONTAINER",
                "Value": "(mackerel|xray)",
              },
            ],
            "Essential": false,
            "Image": "mackerel/mackerel-container-agent:latest",
            "Links": Array [],
            "LinuxParameters": Object {
              "Capabilities": Object {
                "Add": Array [],
                "Drop": Array [],
              },
              "Devices": Array [],
              "Tmpfs": Array [],
            },
            "MemoryReservation": 128,
            "MountPoints": Array [
              Object {
                "ContainerPath": "/host/sys/fs/cgroup",
                "ReadOnly": true,
                "SourceVolume": "cgroup",
              },
              Object {
                "ContainerPath": "/var/run/docker.sock",
                "ReadOnly": true,
                "SourceVolume": "docker_sock",
              },
            ],
            "Name": "mackerel-agent",
            "PortMappings": Array [],
            "Ulimits": Array [],
            "VolumesFrom": Array [],
          },
        ],
        "Family": "TaskDefinition",
        "NetworkMode": "bridge",
        "PlacementConstraints": Array [],
        "RequiresCompatibilities": Array [
          "EC2",
        ],
        "TaskRoleArn": Object {
          "Fn::GetAtt": Array [
            "TaskDefinitionTaskRoleFD40A61D",
            "Arn",
          ],
        },
        "Volumes": Array [
          Object {
            "Host": Object {
              "SourcePath": "/cgroup",
            },
            "Name": "cgroup",
          },
          Object {
            "Host": Object {
              "SourcePath": "/var/run/docker.sock",
            },
            "Name": "docker_sock",
          },
        ],
      },
      "Type": "AWS::ECS::TaskDefinition",
    },
    "TaskDefinitionTaskRoleFD40A61D": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;

exports[`addMackerelContainerAgent specify logGroup 1`] = `
Object {
  "Resources": Object {
    "LogGroupF5B46931": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "RetentionInDays": 731,
      },
      "Type": "AWS::Logs::LogGroup",
    },
    "TaskDefinitionB36D86D9": Object {
      "Properties": Object {
        "ContainerDefinitions": Array [
          Object {
            "Environment": Array [
              Object {
                "Name": "MACKEREL_APIKEY",
                "Value": "keep-my-secret",
              },
              Object {
                "Name": "MACKEREL_CONTAINER_PLATFORM",
                "Value": "ecs",
              },
            ],
            "Essential": false,
            "Image": "mackerel/mackerel-container-agent:latest",
            "Links": Array [],
            "LinuxParameters": Object {
              "Capabilities": Object {
                "Add": Array [],
                "Drop": Array [],
              },
              "Devices": Array [],
              "Tmpfs": Array [],
            },
            "LogConfiguration": Object {
              "LogDriver": "awslogs",
              "Options": Object {
                "awslogs-group": Object {
                  "Ref": "LogGroupF5B46931",
                },
                "awslogs-region": Object {
                  "Ref": "AWS::Region",
                },
                "awslogs-stream-prefix": "mackerel",
              },
            },
            "MemoryReservation": 128,
            "MountPoints": Array [
              Object {
                "ContainerPath": "/host/sys/fs/cgroup",
                "ReadOnly": true,
                "SourceVolume": "cgroup",
              },
              Object {
                "ContainerPath": "/var/run/docker.sock",
                "ReadOnly": true,
                "SourceVolume": "docker_sock",
              },
            ],
            "Name": "mackerel-agent",
            "PortMappings": Array [],
            "Ulimits": Array [],
            "VolumesFrom": Array [],
          },
        ],
        "ExecutionRoleArn": Object {
          "Fn::GetAtt": Array [
            "TaskDefinitionExecutionRole8D61C2FB",
            "Arn",
          ],
        },
        "Family": "TaskDefinition",
        "NetworkMode": "bridge",
        "PlacementConstraints": Array [],
        "RequiresCompatibilities": Array [
          "EC2",
        ],
        "TaskRoleArn": Object {
          "Fn::GetAtt": Array [
            "TaskDefinitionTaskRoleFD40A61D",
            "Arn",
          ],
        },
        "Volumes": Array [
          Object {
            "Host": Object {
              "SourcePath": "/cgroup",
            },
            "Name": "cgroup",
          },
          Object {
            "Host": Object {
              "SourcePath": "/var/run/docker.sock",
            },
            "Name": "docker_sock",
          },
        ],
      },
      "Type": "AWS::ECS::TaskDefinition",
    },
    "TaskDefinitionExecutionRole8D61C2FB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "TaskDefinitionExecutionRoleDefaultPolicy1F3406F5": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "LogGroupF5B46931",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TaskDefinitionExecutionRoleDefaultPolicy1F3406F5",
        "Roles": Array [
          Object {
            "Ref": "TaskDefinitionExecutionRole8D61C2FB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TaskDefinitionTaskRoleFD40A61D": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;

exports[`addMackerelContainerAgent with roles 1`] = `
Object {
  "Resources": Object {
    "TaskDefinitionB36D86D9": Object {
      "Properties": Object {
        "ContainerDefinitions": Array [
          Object {
            "Environment": Array [
              Object {
                "Name": "MACKEREL_APIKEY",
                "Value": "keep-my-secret",
              },
              Object {
                "Name": "MACKEREL_CONTAINER_PLATFORM",
                "Value": "ecs",
              },
              Object {
                "Name": "MACKEREL_ROLES",
                "Value": "My-service:db,My-service:proxy",
              },
            ],
            "Essential": false,
            "Image": "mackerel/mackerel-container-agent:latest",
            "Links": Array [],
            "LinuxParameters": Object {
              "Capabilities": Object {
                "Add": Array [],
                "Drop": Array [],
              },
              "Devices": Array [],
              "Tmpfs": Array [],
            },
            "MemoryReservation": 128,
            "MountPoints": Array [
              Object {
                "ContainerPath": "/host/sys/fs/cgroup",
                "ReadOnly": true,
                "SourceVolume": "cgroup",
              },
              Object {
                "ContainerPath": "/var/run/docker.sock",
                "ReadOnly": true,
                "SourceVolume": "docker_sock",
              },
            ],
            "Name": "mackerel-agent",
            "PortMappings": Array [],
            "Ulimits": Array [],
            "VolumesFrom": Array [],
          },
        ],
        "Family": "TaskDefinition",
        "NetworkMode": "bridge",
        "PlacementConstraints": Array [],
        "RequiresCompatibilities": Array [
          "EC2",
        ],
        "TaskRoleArn": Object {
          "Fn::GetAtt": Array [
            "TaskDefinitionTaskRoleFD40A61D",
            "Arn",
          ],
        },
        "Volumes": Array [
          Object {
            "Host": Object {
              "SourcePath": "/cgroup",
            },
            "Name": "cgroup",
          },
          Object {
            "Host": Object {
              "SourcePath": "/var/run/docker.sock",
            },
            "Name": "docker_sock",
          },
        ],
      },
      "Type": "AWS::ECS::TaskDefinition",
    },
    "TaskDefinitionTaskRoleFD40A61D": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
